lib/evmone/execution.cpp

```diff
@@ -4,26 +4,139 @@
 
 #include "execution.hpp"
 #include "analysis.hpp"
+#include <chrono>
+#include <iostream>
 #include <memory>
 
 namespace evmone
 {
-evmc_result execute(evmc_vm* /*unused*/, const evmc_host_interface* host, evmc_host_context* ctx,
-    evmc_revision rev, const evmc_message* msg, const uint8_t* code, size_t code_size) noexcept
+
+evmc_result execute_measure_collective_time(const evmc_host_interface* host,
+    evmc_host_context* ctx, evmc_revision rev, const evmc_message* msg, const uint8_t* code, size_t code_size) noexcept
 {
     auto analysis = analyze(rev, code, code_size);
 
     auto state = std::make_unique<execution_state>(*msg, rev, *host, ctx, code, code_size);
     state->analysis = &analysis;
 
     const auto* instr = &state->analysis->instrs[0];
+    std::chrono::steady_clock::time_point start_time, end_time;
+    start_time = std::chrono::steady_clock::now();
     while (instr != nullptr)
         instr = instr->fn(instr, *state);
+    end_time = std::chrono::steady_clock::now();
+    std::chrono::nanoseconds elapsed_nanoseconds = end_time - start_time;
+    std::cout << "Collective time: " << elapsed_nanoseconds.count() << std::endl;
+    const auto gas_left =
+        (state->status == EVMC_SUCCESS || state->status == EVMC_REVERT) ? state->gas_left : 0;
+
+    return evmc::make_result(
+        state->status, gas_left, &state->memory[state->output_offset], state->output_size);
+}
+
+evmc_result execute_measure_each_time(const evmc_host_interface* host, evmc_host_context* ctx, evmc_revision rev,
+    const evmc_message* msg, const uint8_t* code, size_t code_size, unsigned int run_id) noexcept
+{
+    auto analysis = analyze(rev, code, code_size);
+
+    auto state = std::make_unique<execution_state>(*msg, rev, *host, ctx, code, code_size);
+    state->analysis = &analysis;
+
+    const auto* instr = &state->analysis->instrs[0];
+    unsigned int instruction_counter = 0;
+    while (instr != nullptr) {
+        std::chrono::steady_clock::time_point start_time, end_time;
+        start_time = std::chrono::steady_clock::now();
```

Comment:
I also tried `__rdtsc()` here. This works and seems to have lower overhead and higher variation. I believe this needs separate research together with `gotsc` implementation.